src/frontc/cabs2cil.ml

```diff
@@ -4847,6 +4917,31 @@ and doExp (asconst: bool)   (* This expression is used as a constant *)
 
     | A.EXPR_PATTERN _ -> E.s (E.bug "EXPR_PATTERN in cabs2cil input")
 
+    | A.GENERIC (expr, lst) -> 
+        let get_exp (_, e, _) = e in
+        let exp = get_exp (doExp false (stripParenLocal expr) (AExp None)) in
+        let cil_list = 
+          let rec make_cil_list cabs_l cil_l = match cabs_l with 
+          | [] -> cil_l
+          | (s, e) :: rest -> make_cil_list rest (((doOnlyType s JUSTBASE),get_exp (doExp false (stripParenLocal e) (AExp None))) :: cil_l) 
+          in
+          make_cil_list lst [] 
+        in
+        let exp_typ = typeOf exp in
+        let default_typ = ref voidType in
+        let typ = 
+          let rec get_typ lst = match lst with 
+            | (TDefault, e) :: rest -> default_typ := typeOf e; get_typ rest
+            | (t, e) :: rest -> if t = exp_typ then typeOf e else get_typ rest
+            | [] -> 
+              if !default_typ = voidType then 
+                E.s(error "Controlling expression of generic selection is not compatible with any association")
+              else !default_typ
+          in
+          get_typ cil_list
+        in
+        finishExp empty (Generic(exp, (cil_list))) typ
```

Comment:
Since the type switch case is already resolved there just to return the correct type, there's no need to keep the `Generic` expression, but we could also just return the expression for that particular case. Then `Generic` doesn't have to exist in CIL AST at all (it's resolved by type during the conversion here). Moreover, Goblint doesn't have to do anything extra to handle `Generic` everywhere as the logic would always be the same.

Reply:
yes, generic should be compiled away!